### PR TITLE
Use preconditioner if sparsity is larger than an empirical threshold

### PR DESCRIPTION
**Pull request by hwpang**
I tested the performance of preconditioned GMRES compared against the default dense solver using 8 chemical kinetic mechanisms with a range of number of species and Jacobian sparsity, and added code to switch the recommended solver from using the default dense linear solver to the preconditioned GMRES linear solver based on an empirical threshold of Jacobian sparsity.

Mechanisms are mostly simulated using ConstantVDomain at 1000 K, 10 bar and a stoichiometric ratio with air. The psetup and prec function code is heavily based on the tutorial in https://diffeq.sciml.ai/stable/tutorials/advanced_ode_example/. It was found that 

- RMG-crafted mechanisms are much denser than similar-sized literature mechanisms. For example, below shows the Jacobian sparsity pattern of RMG-crafted PME mechanism and n-heptane combustion mechanism by Curran et al.

**PME**
image 
**Curran**
image

- As the result, using the sparsity of Jacobian to determine whether to solve the using the default dense solver or the preconditioned sparse solver is more useful than using the number of species in the mechanism. A range of taus from 1e-6 to 1e-1 was used to investigate tau sensitivity. No correlation between the best tau and the number of species or the sparsity of Jacobian was observed. A default tau = 1e-3 performs well. An empirical threshold of 80% sparsity to switch from dense solver to sparse preconditioned solver is selected based on results shown in figure below.

image

### CODE REVIEW

File: src/Reactor.jl [hunk 1]
```diff
@@ -13,13 +16,76 @@ struct Reactor{D,Q} <: AbstractReactor
...
+     sparsity = 1.0 - length(J.nzval)/(y0length*y0length)
+     
+     # preconditioner caches
+     W = spzeros(y0length,y0length)
+     jacy!(W,y0,p,0.0)
+     @. W.nzval = -1.0*W.nzval
+     idxs = diagind(W)
+     @inbounds @views @. W[idxs] = W[idxs] + 1
+     prectmp = ilu(W, τ = tau)
+     preccache = Ref(prectmp)
```
Comment (mjohnson541) [inline src/Reactor.jl:87]: Does this get used?

File: src/Reactor.jl [hunk 2]
```diff
@@ -13,13 +16,76 @@ struct Reactor{D,Q} <: AbstractReactor
...
+        du: the current value of f(u,p,t)
+        gamma: the gamma of W = M - gamma*J
+        delta: the iterative method tolerance
+        lr: a flag for whether lr=1 (left) or lr=2 (right) preconditioning
+        preccache: preconditioner cache
+        """
+        ldiv!(z,preccache[],r)
+    end
+    
+    psetup(p::T1, t::T2, u::T3, du::T4, jok::Bool, jcurPtr::T5, gamma::T6) where {T1,T2,T3,T4,T5,T6} = _psetup(p, t, u, du, jok, jcurPtr, gamma, jacy!, W::SparseMatrixCSC{Float64, Int64}, preccache::Base.RefValue{IncompleteLU.ILUFactorization{Float64, Int64}}, tau::Float64)
```
Comment (mjohnson541) [inline src/Reactor.jl:70]: Can we add all the functions necessary to the Reactor object so people can use these with whatever solvers they want, not just our recommended one?

**Review by mjohnson541 — COMMENTED** [review]
Awesome! I'm a little bit confused how tau gets to where its needed, but apart from that just two questions.

File: src/Reactor.jl (hunk 1 above)
Comment (hwpang) [reply src/Reactor.jl:87]: prectmp doesn't but preccache gets used in `_prec`

**Comment by hwpang** [comment]
`tau` gets fed in from this one line function `psetup(p::T1, t::T2, u::T3, du::T4, jok::Bool, jcurPtr::T5, gamma::T6) where {T1,T2,T3,T4,T5,T6} = _psetup(p, t, u, du, jok, jcurPtr, gamma, jacy!, W::SparseMatrixCSC{Float64, Int64}, preccache::Base.RefValue{IncompleteLU.ILUFactorization{Float64, Int64}}, tau::Float64)`

File: src/Reactor.jl (hunk 2 above)
Comment (hwpang) [reply src/Reactor.jl:72]: Done

File: src/Reactor.jl [hunk 3]
```diff
@@ -9,11 +9,13 @@ using SparseArrays
...
 export AbstractReactor
 
-struct Reactor{D,Q} <: AbstractReactor
+struct Reactor{D,Q,F1,F2} <: AbstractReactor
     domain::D
     ode::ODEProblem
     recommendedsolver::Q
     forwardsensitivities::Bool
+    prec::F1 #function to calculate preconditioner
+    psetup::F2 #function to compute preconditioner \ residue
```
Comment (mjohnson541) [inline src/Reactor.jl:18]: Can we add all the functions necessary to do the preconditioner for the julia solvers?

**Comment by hwpang** [comment]
New push changes names of the functions to distinguish whether the functions are for Sundials or Julia solvers. All preconditioner functions are appended to Reactor.

**Comment by hwpang** [comment]
New push add preconditioner functions to both single and multidomain Reactor attributes, and move preconditioner functions outside of Reactor function.

**Comment by hwpang** [comment]
New push renames static parameter W from the function typing to W1 to avoid conflicts with the local variable W.

**Comment by hwpang** [comment]
@mjohnson541 All tests have passed!

**Pull request merged**